src/protocols/memory/ddr4.stanza

```diff
@@ -0,0 +1,453 @@
+    default => 0.0 +/- 1.00e-12
+  doc: \<DOC>
+  CK.P to DQS.P Inter-signal Timing Skew Constraint in Seconds
+  <DOC>
+  skew-ck-dqs:Toleranced with :
+    default => min-typ-max(-85.0e-12 0.0 935.0e-12)
+  doc: \<DOC>
+  DQ to DQS Inter-pair Timing Skew Constraint in Seconds
+  <DOC>
+  skew-dq-dqs:Toleranced with :
+    default => 0.0 +/- 3.50e-12
+  doc: \<DOC>
+  DM_n to DQS.P Timing Skew Constraint in Seconds
+  <DOC>
+  skew-dm-dqs:Toleranced with:
+    default => 0.0 +/- 3.50e-12 ; Same as DQ timing
+  doc: \<DOC>
+  CMD/ADDR/CTRL to CK Inter-pair Timing Skew Constraint in Seconds
+  <DOC>
+  skew-cmd-addr-ctrl-ck:Toleranced with :
+    default => 0.0 +/- 20.0e-12
+  doc: \<DOC>
+  CMD, ADDR, CTRL (w/o CK) Intra-group Timing Skew Constraint in Seconds
+  <DOC>
+  skew-cmd-addr-ctrl:Toleranced  with:
+    default =>  0.0 +/- 10.0e-12
+  doc: \<DOC>
+  Diff-Pair Max Loss Limit Constraint in dB
+  <DOC>
+  loss:Double with:
+    ensure => ensure-positive!,
+    default => 5.0
+  doc: \<DOC>
+  Differential Routing Structure for CK  [90.0 +/- (5 %)]
+  <DOC>
+  diff-ck-rs:DifferentialRoutingStructure
+  doc: \<DOC>
+  Differential Routing Structure for DQS [100.0 +/- (5 %)]
+  <DOC>
+  diff-dqs-rs:DifferentialRoutingStructure
+  doc: \<DOC>
+  Single-Ended Routing Structure for each DQ signal [50.0 +/- (5 %)]
+  <DOC>
+  se-dq-rs:RoutingStructure
+  doc: \<DOC>
+  Single-Ended Routing Structure for all other SE signal [45.0 +/- (5 %)]
+  <DOC>
+  se-rs:RoutingStructure
+
+with:
+  keyword-constructor => true
+  constructor => #DDR4-Constraint
+
+
+doc: \<DOC>
+Constructor for the DDR4 Link Constraint
+@param ck-rs Differential Routing Structure constraints for CK signals
+@param dqs-rs Differential Routing Structure constraints for DQS signals
+@param dq-rs Routing Structure constraints for DQ signals
+@param rs Routing Structure constraints for remaining signals (CMD/CTRL/ADDR)
+<DOC>
+public defn DDR4-Constraint (
+    width:DDR4-width
+    topology:DDR4-topology
+    rank:DDR4-rank
+    diff-ck-rs:DifferentialRoutingStructure
+    diff-dqs-rs:DifferentialRoutingStructure
+    se-dq-rs:RoutingStructure
+    se-rs:RoutingStructure ) -> DDR4-Constraint:
+  DDR4-Constraint(
+    width = width
+    topology = topology
+    rank = rank
+    diff-ck-rs = diff-ck-rs
+    diff-dqs-rs = diff-dqs-rs
+    se-dq-rs = se-dq-rs
+    se-rs = se-rs
+  )
+
+doc: \<DOC>
+Constrain a DDR4 Link assuming that it is a discrete link (not DIMM)
+
+@param cst Constraint Object
+@param src Source End Point - must be of `ddr4-b` type
+@param dst Destination End Point - must be of `ddr4-b` type and match
+the parameterization of `src`, including width.
+<DOC>
+public defmethod constrain (cst:DDR4-Constraint, src:JITXObject, dst:JITXObject) -> False :
+  inside pcb-module:
+
+    ; val iwidth = width-to-int(width(cst))
+    val ilane  = width-to-lane(width(cst))
+    val irank  = rank-to-int(rank(cst))
+    ; val topo   = topology(cst) ; not sure how to use this atm
```

Comment:
I think there is value in allowing the user to create this statement at the top level. We can probably just remove this.

Reply:
Done. This might need to be amended when we try to constrain multi-rank architectures.